File: kakao/2020/kakao2020_2.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;

bool correct(string p)
{
    int cnt=0;
    for(int i=0;i<p.size();i++)
    {
        if(p[i]=='(')
            cnt++;
        else
        {
            if(!cnt) return false;
            cnt--;
        }
    }
    return true;
}

string solution(string p)
{
    string u,v;
    int rcnt=0,lcnt=0;
    if(p=="") return p;
    for(int i=0;i<p.size();i++)
    {
        if(p[i]=='(') lcnt++;
        else rcnt++;
        if(rcnt==lcnt)
        {
            u=p.substr(0,i+1);
            v=p.substr(i+1);
            break;
        }
    }
    if(correct(u)) return u+solution(v);
    else
    {
        string answer = "";
        answer+="("+solution(v)+")";
        string reversed=u.substr(1,u.length()-2);
        for(int i=0;i<reversed.size();i++)
        {
            if(reversed[i]=='(') answer+=")";
            else answer+="(";
        }
        return answer;
    }
}

File: kakao/2020/kakao2020_2.py
def chk(s):
    cnt=0
    for i in s:
        if i=='(':cnt+=1
        else:
            if cnt==0:return 0
            cnt-=1
    return 1

def solution(p):
    if not p:return p
    rcnt=lcnt=0;u='';v=''
    for i in range(len(p)):
        if p[i]=='(':lcnt+=1
        else:rcnt+=1
        if lcnt==rcnt:u=p[:i+1];v=p[i+1:];break
    if chk(u):return u+solution(v)
    else:
        answer='('+solution(v)+')'
        for i in u[1:-1]:
            if i=='(':answer+=')'
            else:answer+='('
    return answer